plugins/3.mediaPlayer.ts

```diff
@@ -44,65 +46,119 @@ export default defineNuxtPlugin((nuxtApp) => {
   // Good to know when writing tests: https://github.com/jsdom/jsdom/issues/2155#issuecomment-366703395
   let activeMedia: HTMLAudioElement | undefined;
 
+  const loading = ref(false);
   const paused = ref(true);
   const ended = ref(false);
   const currentTrack: Ref<TrackModel | undefined> = ref(undefined);
 
+  const queue: Ref<TrackModel[]> = ref([]);
+  const currentQueueIndex: Ref<number> = ref(0);
+
   const playerStatus = computed(() => {
     if (paused.value) return MediaPlayerStatus.Paused;
     if (ended.value) return MediaPlayerStatus.Stopped;
     return MediaPlayerStatus.Playing;
   });
 
+  function next() {
+    if (queue.value.length > currentQueueIndex.value + 1) {
+      setCurrentTrack(queue.value[currentQueueIndex.value + 1]);
+    }
+  }
+
+  function previous() {
+    if (currentQueueIndex.value > 0) {
+      setCurrentTrack(queue.value[currentQueueIndex.value - 1]);
+    }
+  }
+
+  function clearCurrentTrack() {
+    activeMedia?.pause();
+    activeMedia = undefined;
+    currentTrack.value = undefined;
+    paused.value = true;
+    ended.value = false;
+  }
+
+  function setCurrentTrack(track?: TrackModel) {
+    if (!track) return;
+
+    activeMedia?.pause();
+
+    activeMedia = new Audio(
+      authorizedUrl(track.media?.[0]?.files?.[0]?.url || "", authToken.value)
+    );
+    activeMedia.autoplay = true;
+    loading.value = true;
+    currentTrack.value = track;
+    paused.value = true;
+    ended.value = false;
+
+    // Update queue index if track is in queue
+    // else clear queue and index
+    const index = queue.value.findIndex((t) => t.id === track.id);
+    if (index !== -1) {
+      currentQueueIndex.value = index;
+    } else {
+      queue.value = [];
+      currentQueueIndex.value = 0;
+    }
+
+    activeMedia.addEventListener("pause", () => {
+      paused.value = true;
+    });
+    activeMedia.addEventListener("loadstart", () => {
+      if (activeMedia?.autoplay) {
+        paused.value = false;
+        ended.value = false;
+      }
+      loading.value = false;
+    });
+    activeMedia.addEventListener("play", () => {
+      paused.value = false;
+      ended.value = false;
+    });
+    activeMedia.addEventListener("playing", () => {
+      paused.value = false;
+    });
+    activeMedia.addEventListener("ended", () => {
+      ended.value = true;
+      // Play next track if there is one
+      next();
+    });
+  }
+
+  function addTrackToQueue(track: TrackModel) {
+    // Add track if not already in queue
+    // else move track to end of queue only if it is before the current track
+    const index = queue.value.findIndex((t) => t.id === track.id);
+    if (index === -1) {
+      queue.value.push(track);
+    } else if (index < currentQueueIndex.value) {
+      queue.value.splice(index, 1);
+      queue.value.push(track);
+    }
+
+    // Play if currently paused and no track is playing
```

Comment:
I guess I'll have to rip this apart in another PR... I can just come up with all to many scenarios where this wouldn't be what a user would expect to happen... E.g. when playing some songs but clicking on a song of an album ...